doc/lha.1
.TH lha 1
.SH NAME
lha \- compression tool for .lzh archive files.
.SH SYNOPSIS
.B lha
.RB [ - ]{ lvtxep [ q { \f[I]num\f[] }][ finv ]}[ w= < \f[I]dir\f[] >]
.I archive_file
.RI [ "file ..." ]
.SH DESCRIPTION
.PP
.B lha
is a tool for extracting .lzh archive files. It also supports variants
of the .lzh archive, such as .lzs and .pma.
.PP
This version of the lha tool is part of Lhasa, a free implementation
of the .lzh format.
.PP
.SH COMMAND SYNTAX
The lha tool has an unusual command syntax, compared to most other
Unix commands. The first parameter to the program specifies the command
to perform and all additional options. The second parameter specifies
the path to the archive file to operate on. Following this is a list
of wildcard patterns to match against the filenames of the archived
files.
.PP
The first character of the command parameter specifies the command to
perform, which is one of the following:
.TP
\fB-l\fR
List contents of the specified archive.
.TP
\fB-v\fR
Verbosely list contents of the specified archive.
.TP
\fB-t\fR
Test the integrity of the specified archive: decompress its contents and
check the CRC.
.TP
\fB-e\fR or \fB-x\fR
Extract archive. Files are extracted to the current working directory
unless the 'w' option is specified.
.TP
\fB-p\fR
Extract archive, sending decompressed files to stdout rather than
writing them to the filesystem as actual files. This is useful when used
as part of a shell pipeline.
.PP
.SH OPTIONS
The remainder of the command parameter is used to specify additional
options:
.TP
\fBq[012]\fR
Quiet mode. Higher numbers suppress more output. Level 0 is normal
operation. If no number is specified, full suppression (level 2)
is used. The quiet option also turns on the force overwrite option
('f').
.TP
\fBf\fR
Force overwrite of existing files: do not prompt.
.TP
\fBi\fR
Ignore paths of archived files: extract all archived files to the
same directory, ignoring subdirectories.
.TP
\fBn\fR
Do not perform any actual operations: instead, perform a dry run of
the requested operation and describe what would have been done on
standard output.
.TP
\fBv\fR
Verbose mode: causes extra information to be written to standard
output.
.TP
\fBw=dir\fR
Specify destination directory for extracting files. This must be
the last option of the first parameter.
.SH LIST OUTPUT FORMAT
Lhasa inherits its list output format from the original Unix port of lha, and
retains the same format for compatibility, since some tools parse the output.
Four different output formats are supported, depending on the command line
arguments:
.TP
\fBlha -l filename.lzh\fR
Shows: Unix permissions / OS type, UID/GID, uncompressed size, compression
ratio, modification time, filename.
.TP
\fBlha -lv filename.lzh\fR
Two line output. First line shows the filename, second line shows: Unix
permissions / OS type, UID/GID, uncompressed size, compression ratio,
modification time, filename, header level.
.TP
\fBlha -v filename.lzh\fR
Shows: Unix permissions / OS type, UID/GID, compressed size, uncompressed size,
compression ratio, compression format, checksum, modification time, filename.
.TP
\fBlha -vv filename.lzh\fR
Two line output. First line shows the filename, second line shows: Unix
permissions / OS type, UID/GID, compressed size, uncompressed size, compression
ratio, compression format, checksum, long form modification time, header
level.
.PP
The following are detailed descriptions of each of these fields:
.TP
\fBUnix permissions / OS type (PERMSSN)\fR
For files compressed on a Unix system, this lists the permissions in the same
format that \fBls\fR(1) uses when it is invoked with the \fB-l\fR option. If
the file was not created on a Unix system, the OS type is instead displayed in
square brackets. Examples are "[MS-DOS]" and "[MacOS]"; the OS name never
contains a space. If the file was created on Microware OS-9, the permissions
for that system will also be shown.
.TP
\fBUnix user ID / group IDs (UID/GID)\fR
For files compressed on a Unix system, this lists the user and group IDs of
the file, similar to the output seen from \fBls\fR(1) when it is invoked with
the \fB-ln\fR option. If the file was not created on a Unix system,
"*****/*****" is shown; see \fBLIST FORMAT DIFFERENCES\fR below.
.TP
\fBCompressed size (PACKED)\fR
Size of the file in bytes as it is stored inside the archive (not including
the header, which contains metadata about the file). If an uncompressed format
like \fB-lh0-\fR is used, this will be equal to SIZE.
.TP
\fBUncompressed size (SIZE)\fR
Size of the file in bytes before it was compressed.
.TP
\fBFile compression ratio (RATIO)\fR
Compressed size divided by uncompressed size. This measures the effectiveness
of the compression format; the smaller the percentage, the better. If an
uncompressed format like \fB-lh0-\fR is used, this will be equal to 100%. For
directories, "******" is shown.
Note that the value here is backwards relative to \fBunzip\fR(1), which instead
shows the percentage of the original file size reduced.
.TP
\fBCompression format (METHOD)\fR
The compression format used to store this file. See the \fBCOMPRESSION
FORMATS\fR section below.
.TP
\fBChecksum (CRC)\fR
CRC16 checksum of the file's contents in hexadecimal format. Warning: while
this is useful for casually checking file integrity, it is only a 16-bit
checksum and should not be considered reliable for protecting files larger than
a few kilobytes.
.TP
\fBModification time (STAMP)\fR
Date and time that the file was last modified before it was stored in the
archive. If the file doesn't have a timestamp stored, nothing is shown.
The output can be in one of two forms: if the date is within the past six
months, the month, day and time are shown; an example is "Aug 24 12:38". For
older files, the month, day and year are shown; an example is "Aug 24 2020".
.TP
\fBLong form modification time (STAMP)\fR
This is only used when listing in the \fB-vv\fR format. The full date and
timestamp are listed in numeric form, close to ISO 8601 format (though not
quite conformant with that standard). An example is "2020-08-24 12:38:25".
.TP
\fBHeader level (LV)\fR
Lha header format used to store this file. Supported formats are 0-3.
.SH LIST FORMAT DIFFERENCES
The list mode output from the original Unix lha tool was not designed to be
machine-readable. Since some tools do try to parse the list output, Lhasa's
list output has been deliberately tweaked slightly to make the output more
consistent and easier to parse.
.PP
Here are the main differences:
.IP \(bu
If a file was not created on a Unix system, a spacer word is shown in the
UID/GID column. The original lha just showed whitespace.
.IP \(bu
The OS type for archives generated by MacLHA is listed as "MacOS" (no space).
The original lha showed "Mac OS".
.PP
Note that in general, the approach of parsing text output is inherently fragile
and prone to bugs. It's recommended that you instead use the proper API
(liblhasa).
.SH COMPRESSION FORMATS
The following gives some basic description of the various different supported
compression formats (as listed when using the list command documented above).
All formats are variants on the LZSS compression algorithm.
.TP
\fB\-lz4\-\fR
Uncompressed (stored) data, as used by the original LArc tool. This is not
related to the LZ4 compression algorithm of the same name.
.TP
\fB\-lz5\-\fR, \fB\-lzs\-\fR
Compression formats introduced with the original LArc tool.
\-lzs\- uses a 2KiB sliding window while \-lz5\- uses a 4KiB window.
.TP
\fB\-lhd\-\fR
Directory entry. No data is stored, only headers.
.TP
\fB\-lh0\-\fR
Uncompressed (stored) data, as used in LHarc and LHa.
.TP
\fB\-lh1\-\fR
Compression algorithm introduced with LHarc. This uses a 4KiB sliding window
with dynamic Huffman encoding.
.TP
\fB\-lh4\-\fR, \fB\-lh5\-\fR, \fB\-lh6\-\fR, \fB\-lh7\-\fR
New algorithm introduced with LHa (aka LHarc v2.0). The original versions
(\-lh4\- and \-lh5\-) used 8KiB and 16KiB sliding window sizes, respectively.
Later versions introduced new variants with larger window sizes: \-lh6\-
(64KiB) and \-lh7\- (128KiB).
.TP
\fB\-lhx\-\fR
Variant on the \-lh4\- format listed above introduced by UNLHA32.dll that
extends the window size to 1MiB.
.TP
\fB\-lk7\-\fR
Variant on the \-lh4\- format listed above that was introduced by the LHark
tool (a fork of LHa). This is actually named \-lh7\- within the archive files,
but Lhasa renames it internally to avoid clashing with the other algorithm of
the same name. Uses a 128KiB sliding window, with some minor improvement to the
algorithm.
.TP
\fB\-pm0\-\fR
Uncompressed (stored) data, as used by the PMarc tool.
.TP
\fB\-pm1\-\fR
Algorithm used by version 1 of the PMarc tool.
.TP
\fB\-pm2\-\fR
Algorithm used by version 2 of the PMarc tool.
.PP
.SH UNSUPPORTED FORMATS
The following formats are \fBnot\fR currently supported by Lhasa (but possibly
will be in the future):
.TP
\fB\-lh2\-\fR, \fB\-lh3\-\fR
These formats can be decompressed by LHa v2.0 but the tool could not actually
generate archives with these formats; they appear to have only been supported
in beta versions before the v2.0 release. As a result, few if any examples
exist in the wild.
.TP
\fB\-lh8\-\fR, \fB\-lh9\-\fR, \fB\-lha\-\fR, \fB\-lhb\-\fR, \fB\-lhc\-\fR, \fB\-lhe\-\fR
Extensions of the \-lh4\- format to even larger window sizes.
.TP
\fB\-ll0\-\fR, \fB\-ll1\-\fR
Generated by ThunderSoft's PAKLEO utility (.pll).
.TP
\fB\-sw0\-\fR, \fB\-sw1\-\fR
Used in SourceWare Archival Group's .swg (SWAG packet) format.
.TP
\fB\-lz6\-\fR, \fB\-LD6\-\fR
Generated by LDarc and LDIFF (.lzd).
.TP
\fB\-hf0\-\fR, \fB\-ah0\-\fR, \fB\-ari\-\fR, \fB\-arn\-\fR, \fB\-arw\-\fR
Generated by the Micrognosis Compression Archiver utility.
.TP
\fB\-pc1\-\fR
Format generated by the PopCom! compression utility for CP/M.
.PP
If you encounter examples of these in the wild, please get in touch.
.SH EXAMPLES
Here are some examples for how to invoke the program:
.TP
lha -v foobar.lzs
List the contents of the file \fBfoobar.lzs\fR (producing verbose output).
.TP
lha -xf foobar.exe
Extract the contents of a self-extracting archive file named \fBfoobar.exe\fR
to the current directory, overwriting existing files with the same names if
they exist.
.TP
lha -xqw=/tmp foobar.lzh
Extract the contents of \fBfoobar.lzh\fR to \fB/tmp\fR, overwriting any
existing files found there, and suppressing normal output (similar to
how other Unix tools such as \fBcp\fR(1) or \fBtar\fR(1) act silently
by default).
.SH WWW
.UR https://lhasa.soulsphere.org/
https://lhasa.soulsphere.org/
.UE
.SH BUG REPORTS
Bugs can be reported to the GitHub issue tracker:
.UR https://github.com/fragglet/lhasa
https://github.com/fragglet/lhasa
.UE
.SH SEE ALSO
\fBunzip\fR(1), \fBtar\fR(1), \fBgzip\fR(1), \fBbzip2\fR(1),
\fBxz\fR(1), \fBlzip\fR(1)
.SH HISTORY
The .lzh format originated with Kazuhiko Miki's MS\-DOS archive tool,
LArc, using the LZSS algorithm developed by Haruhiko Okumura, and
the .lzs filename extension. The container format was reused for
LHarc, by Haruyasu Yoshizaki (Yoshi), which used a new algorithm
named LZHUF and the .lzh extension. In later versions, LHarc was
renamed to LHA and extended with more effective compression algorithms.
.PP
Versions of the LHA tool were later ported to various different
operating systems, including the Amiga, Atari, MacOS, OS/2 and Unix.
A tool for MSX\-DOS named PMarc reused the container format with a new
compression algorithm (.pma extension).
.PP
The Unix version of the tool was developed by Masaru Oki, Nobutaka
Watazaki and Tsugio Okamoto, but was released under a software
license that does not conform to the Free Software or Open Source
Definitions. Lhasa was developed as a drop\-in replacement that is
Free Software and Open Source.
.PP
The name "LHA" and the .lzh filename extension are a reference to the
compression algorithm: a combination of Lempel–Ziv–Storer–Szymanski
dictionary coding and Huffman encoding. Similar schemes are used in
other compression algorithms such as Deflate (used in gzip/zlib, PNG
image files, and the .zip format).
.SH BUGS
The current version does not allow the creation of new archive files.
.PP
Some obscure compression algorithms are not currently supported (see the
UNSUPPORTED FORMATS section above).
.PP
The tool does not currently do text format conversion for non-ASCII
filenames when listing the contents of archives. Non-ASCII characters
are replaced by a question mark.
.SH AUTHOR
Lhasa was written and is maintained by
.MT fraggle@gmail.com
Simon Howard
.ME .
.SH COPYRIGHT
Copyright \(co 2011-2025 Simon Howard.
.PP
Permission to use, copy, modify, and/or distribute this software
for any purpose with or without fee is hereby granted, provided
that the above copyright notice and this permission notice appear
in all copies.
.PP
THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


NEWS
v0.5.0 (????-??-??):

     * The output from the list subcommands has been tweaked to be more
       consistent. This should help to avoid parsing bugs with
       naively-written programs and scripts that parse the list output.
       This includes various GUI tools.
     * The test suite was expanded to include archives generated by Tascal
       Lha (a Windows CE port); LH/2 (an OS/2 port); ExpLZH; and the port
       of Lha that ships with MorphOS.
     * A workaround was added for a bug with the Amiga port of lha that
       causes some versions to generate malformed archives where
       directories are stored using the -lh0- compression type. Thanks to
       @zzarko, @polluks, @alfredone and @jsummers for their work on
       figuring out the bug.
     * Support was added for the 64-bit file sizes header, allowing huge
       files (>=4GiB) generated by the MorphOS port of lha to be
       extracted correctly.
     * The manual page now includes more detailed information about the
       different list subcommands.
     * The liblhasa headers are now installed into a directory with a name
       that accurately reflects the project's version number.
     * The liblhasa .so version numbers now have meaningful numbers.
     * Some error messages were changed to print filenames safely.
     * Several minor memory leaks were fixed.
     * Extraction of DECLHA self-extracting archives was fixed.
     * Tests can now be run correctly in parallel.
     * The Doxygen search feature was enabled for the documentation pages.
     * Undefined behavior in the BitStreamReader code was fixed (thanks
       @sezero)
     * Some static arrays were made constant (thanks Cameron Cawley).
     * Out-of-tree builds were fixed (thanks Daisuke Fujimura)

v0.4.0 (2023-05-14):

     * The manpage was expanded, with more information about different
       compression formats, example commands, a BUGS section and some
       additional references in the SEE ALSO section.
     * A verbose mode was added to the list output that shows the full file
       timestamp (thanks Dan Fandrich).
     * Support for archives generated by the DOS LHARK tool was added; big
       thanks go to Jason Summers for his tool DEARK and his comprehensive
       writeup on the LHARK format.
     * Some changes were made for compatibility with old versions of
       autotools (thanks Ozkan Sezer).
     * Some tweaks were made to tests to make them work on BSD systems.
     * Segmentation fault on NetBSD/amd64 was fixed when decoding 8-bit
       character filenames (thanks Izumi Tsutsui).
     * Some minor tweaks and fixes were made to the help text and manpage
       (thanks Stefan)
     * Some spelling mistakes were fixed, excess whitespace was trimmed, and
       the website link in the RPC spec file was fixed (thanks a1346054 and
       Jakub Wilk).
     * A potential integer overflow in the lha_decode_*uint32 functions was
       fixed (thanks Jacub Wilk).

v0.3.1 (2016-03-29):

     * This release fixes an integer underflow vulnerability in the
       code for doing LZH level 3 header decodes (TALOS-CAN-0095).
       Thanks go to Marcin Noga and Regina Wilson of Cisco TALOS for
       reporting this vulnerability.

v0.3.0 (2015-04-20):

     * PMarc -pm1- archives that contain truncated compressed data (the
       decompressed length is longer than what can be read from the
       compressed data) now decompress as intended. Certain archives
       in the wild make the assumption that this can be done.
     * LArc -lz5- archives that make use of the initial history buffer
       now decompress correctly.
     * The tests no longer use predictable temporary paths (thanks Jon
       Dowland).
     * Tests were fixed under OS X.

v0.2.0 (2013-08-04):

     * Decompression of archives using the -lhx- file format supported by
       unlha32.dll is now supported (thanks Multi for the patch).
     * The -p (print to stdout) command line option is now supported.
     * The test suite should now run correctly on Windows.

    Bugs fixed:
     * Bug where archives read from pipes (eg. stdin) were not extracted
       beyond the first file in the archive.
     * Output when using the -w (extract directory) option now correctly
       matches the output of Unix lha.

v0.1.0 (2013-03-16):

     * There are now test archives for OS-9 and OS-9/68k (OSK) and a
       workaround for a bug in the OSK lha tool on this platform. OSK level
       0 extended areas are also supported.
     * Extracted files are now written using O_EXCL, which prevents
       malicious symbolic links being used to redirect output.
     * Directory paths containing '..' as a directory name are now
       sanitized, to prevent malicious archives being able to overwrite
       arbitrary files on the filesystem.
     * Symbolic links are now extracted in a safer way, being created as
       dummy files that are overwritten with proper symbolic links at the
       end of extraction. This is the same behavior used by GNU tar to
       prevent malicious use of symbolic links.
     * Automake 1.13 is now properly supported (thanks Jan Engelhardt).
       Processing of archives read from IPC pipes (including stdin) has
       been fixed.

v0.0.7 (2012-06-02):

     * Extraction and listing of Unix symbolic links is now supported.
     * Decompression code for the "old" PMarc archive algorithm (-pm1-) has
       been added.
     * Support has been added for Unix LHA level 0 header extended areas
       (so level 0 archives with Unix metadata are now listed and extracted
       correctly).
     * The Unix permissions field in the list output for directory entries
       has been fixed.
     * The library header files have been fixed so that they can be included
       in C++ code.
     * The LHADecoder interface, for extracting raw compressed data, has been
       added to the public header files.
     * The Unix LHA test archives have been regenerated and improved.

src/lha.c
/*

Copyright (c) 2011, Simon Howard

Permission to use, copy, modify, and/or distribute this software
for any purpose with or without fee is hereby granted, provided
that the above copyright notice and this permission notice appear
in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>

#include <unistd.h>
#include <sys/stat.h>

#include "lha_reader.h"

static void help_page(char *progname)
{
	printf("usage: %s [-]{lv} archive_file\n", progname);

	printf("commands:\n"
	       " l,v List / Verbose List\n");
	exit(-1);
}

static float compression_percent(unsigned int compressed,
                                 unsigned int uncompressed)
{
	float factor;

	if (uncompressed > 0) {
		factor = (float) compressed / uncompressed;
	} else {
		factor = 1.0;
	}

	return factor * 100.0;
}

typedef struct
{
	unsigned int num_files;
	unsigned int compressed_length;
	unsigned int length;
	unsigned int timestamp;
} FileStatistics;

typedef struct
{
	char *name;
	unsigned int width;
	void (*handler)(LHAFileHeader *header);
	void (*footer)(FileStatistics *stats);
} ListColumn;

// Display OS type:

static char *os_type_to_string(uint8_t os_type)
{
	switch (os_type) {
		case LHA_OS_TYPE_MSDOS:
			return "[MS-DOS]";
		case LHA_OS_TYPE_UNIX:
			return "[Unix]";
		case LHA_OS_TYPE_OS2:
			return "[OS/2]";
		case LHA_OS_TYPE_UNKNOWN:
			return "[generic]";
		default:
			return "[unknown]";
	}
}

// File permissions

static void permission_column_print(LHAFileHeader *header)
{
	const char *perms = "drwxrwxrwx";
	unsigned int i;

	if ((header->extra_flags & LHA_FILE_UNIX_PERMS) == 0) {
		printf("%-10s", os_type_to_string(header->os_type));
		return;
	}

	for (i = 0; i < 10; ++i) {
		if (header->unix_perms & (1 << (9 - i))) {
			printf("%c", perms[i]);
		} else {
			printf("-");
		}
	}
}

static void permission_column_footer(FileStatistics *stats)
{
	printf(" Total    ");
}

static ListColumn permission_column = {
	" PERMSSN", 10,
	permission_column_print,
	permission_column_footer
};

// Unix UID/GID

static void unix_uid_gid_column_print(LHAFileHeader *header)
{
	if (header->extra_flags & LHA_FILE_UNIX_UID_GID) {
		printf("%5i/%-5i", header->unix_uid, header->unix_gid);
	} else {
		printf("           ");
	}
}

static void unix_uid_gid_column_footer(FileStatistics *stats)
{
	// The UID/GID column has the total number of files
	// listed below it.

	if (stats->num_files == 1) {
		printf("%5i file ", stats->num_files);
	} else {
		printf("%5i files", stats->num_files);
	}
}

static ListColumn unix_uid_gid_column = {
	" UID  GID", 11,
	unix_uid_gid_column_print,
	unix_uid_gid_column_footer
};

// Compressed file size

static void packed_column_print(LHAFileHeader *header)
{
	printf("%7i", header->compressed_length);
}

static void packed_column_footer(FileStatistics *stats)
{
	printf("%7i", stats->compressed_length);
}

static ListColumn packed_column = {
	" PACKED", 7,
	packed_column_print,
	packed_column_footer
};

// Uncompressed file size

static void size_column_print(LHAFileHeader *header)
{
	printf("%7i", header->length);
}

static void size_column_footer(FileStatistics *stats)
{
	printf("%7i", stats->length);
}

static ListColumn size_column = {
	"   SIZE", 7,
	size_column_print,
	size_column_footer
};

// Compression ratio

static void ratio_column_print(LHAFileHeader *header)
{
	if (!strcmp(header->compress_method, "-lhd-")) {
		printf("******");
	} else {
		printf("%5.1f%%", compression_percent(header->compressed_length,
		                                      header->length));
	}
}

static void ratio_column_footer(FileStatistics *stats)
{
	printf("%5.1f%%", compression_percent(stats->compressed_length,
	                                      stats->length));
}

static ListColumn ratio_column = {
	" RATIO", 6,
	ratio_column_print,
	ratio_column_footer
};

// Compression method and CRC checksum

static void method_crc_column_print(LHAFileHeader *header)
{
	printf("%-5s %04x", header->compress_method, header->crc);
}

static ListColumn method_crc_column = {
	"METHOD CRC", 10,
	method_crc_column_print
};

// File timestamp

static void output_timestamp(unsigned int timestamp)
{
	const char *months[] = {
		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
	};
	struct tm *ts;
	time_t tmp;

	if (timestamp == 0) {
		printf("            ");
		return;
	}

	tmp = timestamp;
	ts = localtime(&tmp);

	// Print date:

	printf("%s %2d ", months[ts->tm_mon], ts->tm_mday);

	// If this is an old time (more than 6 months), print the year.
	// For recent timestamps, print the time.

	tmp = time(NULL);

	if (timestamp > tmp - 6 * 30 * 24 * 60 * 60) {
		printf("%02i:%02i", ts->tm_hour, ts->tm_min);
	} else {
		printf(" %04i", ts->tm_year + 1900);
	}
}

static void timestamp_column_print(LHAFileHeader *header)
{
	output_timestamp(header->timestamp);
};

static void timestamp_column_footer(FileStatistics *stats)
{
	output_timestamp(stats->timestamp);
};

static ListColumn timestamp_column = {
	"    STAMP", 12,
	timestamp_column_print,
	timestamp_column_footer
};

// Filename

static void name_column_print(LHAFileHeader *header)
{
	if (header->path != NULL) {
		printf("%s", header->path);
	}
	printf("%s", header->filename);
}

static ListColumn name_column = {
	"       NAME", 20,
	name_column_print
};

static ListColumn short_name_column = {
	"      NAME", 13,
	name_column_print
};

// Print the names of the column headings at the top of the file list.

static void print_list_headings(ListColumn **columns)
{
	unsigned int i, j;

	for (i = 0; columns[i] != NULL; ++i) {
		j = printf("%s", columns[i]->name);

		if (columns[i + 1] != NULL) {
			for (; j < columns[i]->width + 1; ++j) {
				printf(" ");
			}
		}
	}

	printf("\n");
}

// Print separator lines shown at top and bottom of file list.

static void print_list_separators(ListColumn **columns)
{
	unsigned int i, j, len;

	for (i = 0; columns[i] != NULL; ++i) {
		for (j = 0; j < columns[i]->width; ++j) {
			printf("-");
		}

		if (columns[i + 1] != NULL) {
			printf(" ");
		}
	}

	printf("\n");
}

// Print a row in the list corresponding to a file.

static void print_columns(ListColumn **columns, LHAFileHeader *header)
{
	unsigned int i;

	for (i = 0; columns[i] != NULL; ++i) {
		columns[i]->handler(header);

		if (columns[i + 1] != NULL) {
			printf(" ");
		}
	}

	printf("\n");
}

// Print footer information shown at end of list (overall file stats)

static void print_footers(ListColumn **columns, FileStatistics *stats)
{
	unsigned int i, j, len;
	unsigned int num_columns;

	// Work out how many columns there are to print, ignoring trailing
	// columns that have no footer:

	num_columns = 0;

	for (i = 0; columns[i] != NULL; ++i) {
		++num_columns;
	}

	while (num_columns > 0 && columns[num_columns-1]->footer == NULL) {
		--num_columns;
	}

	// Print footers for each column.
	// Some columns do not have footers: fill in with spaces instead.

	for (i = 0; i < num_columns; ++i) {
		if (columns[i]->footer != NULL) {
			columns[i]->footer(stats);
		} else if (columns[i + 1] != NULL) {
			len = strlen(columns[i]->name);

			for (j = 0; j < len; ++j) {
				printf(" ");
			}
		}

		if (i + 1 < num_columns) {
			printf(" ");
		}
	}

	printf("\n");
}

static unsigned int read_file_timestamp(FILE *fstream)
{
	struct stat data;

	if (fstat(fileno(fstream), &data) != 0) {
		return (unsigned int) -1;
	}

	return data.st_mtime;
}

static void list_file_contents(char *filename, ListColumn **columns)
{
	FILE *fstream;
	LHAInputStream *stream;
	LHAReader *reader;
	FileStatistics stats;

	fstream = fopen(filename, "rb");

	if (fstream == NULL) {
		fprintf(stderr, "LHa: Error: %s %s\n",
		                filename, strerror(errno));
		exit(-1);
	}

	print_list_headings(columns);
	print_list_separators(columns);

	stream = lha_input_stream_new(fstream);
	reader = lha_reader_new(stream);
	stats.num_files = 0;
	stats.length = 0;
	stats.compressed_length = 0;
	stats.timestamp = read_file_timestamp(fstream);

	for (;;) {
		LHAFileHeader *header;

		header = lha_reader_next_file(reader);

		if (header == NULL) {
			break;
		}

		print_columns(columns, header);

		++stats.num_files;
		stats.length += header->length;
		stats.compressed_length += header->compressed_length;
	}

	lha_reader_free(reader);
	lha_input_stream_free(stream);

	print_list_separators(columns);
	print_footers(columns, &stats);

	fclose(fstream);
}

static ListColumn *normal_column_headers[] = {
	&permission_column,
	&unix_uid_gid_column,
	&size_column,
	&ratio_column,
	&timestamp_column,
	&name_column,
	NULL
};

static ListColumn *verbose_column_headers[] = {
	&permission_column,
	&unix_uid_gid_column,
	&packed_column,
	&size_column,
	&ratio_column,
	&method_crc_column,
	&timestamp_column,
	&short_name_column,
	NULL
};

int main(int argc, char *argv[])
{
	if (argc < 3) {
		help_page(argv[0]);
	}

	if (!strcmp(argv[1], "l")) {
		list_file_contents(argv[2], normal_column_headers);
	} else if (!strcmp(argv[1], "v")) {
		list_file_contents(argv[2], verbose_column_headers);
	} else {
		help_page(argv[0]);
	}
}

